planner/core/testdata/plan_suite_out.json

```diff
@@ -1420,6 +1420,94 @@
       }
     ]
   },
+  {
+    "Name": "TestPushdownDistinctEnable",
+    "Cases": [
+      {
+        "SQL": "select /*+ HASH_AGG() */ a, count(distinct a) from t;",
+        "Plan": [
+          "Projection_4 1.00 root  test.t.a, Column#5",
+          "└─HashAgg_8 1.00 root  funcs:count(distinct test.t.a)->Column#5, funcs:firstrow(Column#6)->test.t.a",
+          "  └─TableReader_9 1.00 root  data:HashAgg_5",
+          "    └─HashAgg_5 1.00 cop[tikv]  group by:test.t.a, funcs:firstrow(test.t.a)->Column#6",
+          "      └─TableFullScan_7 10000.00 cop[tikv] table:t keep order:false, stats:pseudo"
+        ],
+        "Result": [
+          "1 2"
+        ]
+      },
+      {
+        "SQL": "select /*+ HASH_AGG() */ avg(b), c, avg(b), count(distinct A, B),  count(distinct A), count(distinct c), sum(b) from t group by c;",
+        "Plan": [
+          "Projection_4 8000.00 root  Column#5, test.t.c, Column#5, Column#6, Column#7, Column#8, Column#9",
+          "└─HashAgg_8 8000.00 root  group by:test.t.c, funcs:avg(Column#10, Column#11)->Column#5, funcs:count(distinct test.t.a, test.t.b)->Column#6, funcs:count(distinct test.t.a)->Column#7, funcs:count(distinct test.t.c)->Column#8, funcs:sum(Column#12)->Column#9, funcs:firstrow(test.t.c)->test.t.c",
+          "  └─TableReader_9 8000.00 root  data:HashAgg_5",
+          "    └─HashAgg_5 8000.00 cop[tikv]  group by:test.t.a, test.t.b, test.t.c, funcs:count(test.t.b)->Column#10, funcs:sum(test.t.b)->Column#11, funcs:sum(test.t.b)->Column#12",
+          "      └─TableFullScan_7 10000.00 cop[tikv] table:t keep order:false, stats:pseudo"
+        ],
+        "Result": [
+          "1.0000 1 1.0000 1 1 1 1",
+          "1.3333 3 1.3333 3 2 1 4",
+          "2.0000 <nil> 2.0000 1 1 0 2"
+        ]
+      },
+      {
+        "SQL": "select /*+ STREAM_AGG() */ count(distinct c) from t;",
+        "Plan": [
+          "StreamAgg_12 1.00 root  funcs:count(distinct test.t.c)->Column#5",
+          "└─IndexReader_13 1.00 root  index:StreamAgg_7",
+          "  └─StreamAgg_7 1.00 cop[tikv]  group by:test.t.c, ",
+          "    └─IndexFullScan_11 10000.00 cop[tikv] table:t, index:c(c) keep order:false, stats:pseudo"
```

Comment:
`keep order` should be `true` in order to produce the correct result.

Reply:
It's still `keep order: false` in master(5cf0c3e2d) branch, I think it's another issue.
```
mysql> desc select /*+ STREAM_AGG(), AGG_TO_COP() */ count(c) from t;
+----------------------------+---------+-----------+---------------------+---------------------------------+
| id                         | estRows | task      | access object       | operator info                   |
+----------------------------+---------+-----------+---------------------+---------------------------------+
| StreamAgg_12               | 1.00    | root      |                     | funcs:count(Column#9)->Column#8 |
| └─IndexReader_13           | 1.00    | root      |                     | index:StreamAgg_7               |
|   └─StreamAgg_7            | 1.00    | cop[tikv] |                     | funcs:count(test.t.c)->Column#9 |
|     └─IndexFullScan_11     | 4.00    | cop[tikv] | table:t, index:c(c) | keep order:false, stats:pseudo  |
+----------------------------+---------+-----------+---------------------+---------------------------------+
4 rows in set (0.00 sec)
```

Reply:
> `keep order` should be `true` in order to produce the correct result.

Actually this PR can produce the correct result. It won't affect the property the stream aggregate can provide to its parent, it also won't affect the implicit(not the *explicit*) required property for its child.

If we set `keep order` to `true` when the original `group by` column set size is 0 but the modified `group by` column set has at least 1 column, it can be much faster because we can reduce much more tuples in the partial aggregation. You can tell the differences from the following example:

> It's still `keep order: false` in master(5cf0c3e) branch

master branch does not have this issue: there is group by column but keep order is false.

Reply:
Thanks. I've added some rules to filter out streamAgg in that situation. For example, 
`select count(distinct c) from t` will never use streamAgg when push down option is enabled.
`select /*+ STREAM_AGG() */ count(distinct c) from t` will never push down even push down option is enabled.
`select /*+ STREAM_AGG() */ count(distinct c) from t group by c` will use streamAgg with push down correctly when push down option is enabled.